submission/hammer/main.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Hammer tool sends multiple add-pre-chain requests to Submission proxy at the same time.
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/submission"
+)
+
+// Default number of submissions is explicitly low.
+// After several runs, likely to hit Log rate-limits.
+var (
+	proxyEndpoint = flag.String("proxy_endpoint", "http://localhost:5951/", "Endpoint for HTTP (host:port)")
+	timeout       = flag.Duration("duration", 0*time.Minute, "Time to run continuous flow of submissions")
+	count         = flag.Int("count", 10, "Total number of submissions to execute")
+	qps           = flag.Int("qps", 5, "Number of requests per second")
+)
+
+func main() {
+	wd, _ := os.Getwd()
+	certData, err := ioutil.ReadFile(wd + "/submission/hammer/testdata/precert.der")
+	if err != nil {
+		log.Fatalf("%v\n", err)
+	}
+	interimData, err := ioutil.ReadFile(wd + "/submission/hammer/testdata/interim.der")
+	if err != nil {
+		log.Fatalf("%v\n", err)
+	}
+
+	var req ct.AddChainRequest
+	req.Chain = append(req.Chain, certData)
+	req.Chain = append(req.Chain, interimData)
+
+	postBody, err := json.Marshal(req)
+	if err != nil {
+		log.Fatalf("%v\n", err)
+	}
+
+	// Submission runs until timeout or count is reached. If any of those flags
+	// is set to 0, only other flag value is used as restriction.
+	ctx, cancel := context.WithCancel(context.Background())
+	if *timeout > 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), *timeout)
+	} else if *count <= 0 {
+		// Both restrictions set to 0. Nothing to run.
```

Comment:
Log something here to make it clear why the tool does nothing.

Reply:
Done